default/python/AI/Trait/Character.py

```diff
@@ -0,0 +1,687 @@
+    # should be removed as a trait.
+
+    # TODO: Remove EmpireIDTrait.
+    # Empire id is a game mechanic.  It is not something that a player
+    # can describe, "Look the 'Continuum' is behaving like a 1 modulo 2 character."
+
+    def __init__(self, empire_id, aggression):
+        print "EmpireIDTrait initialized."
+        self.id = empire_id
+        self.aggression = aggression  # TODO remove when old research style get_research_index is removed
+
+    @property
+    def key(self):
+        return self.id % 2
+
+    def _modulo_two_choice(self, alternatives):
+        if not alternatives:
+            return None
+        if len(alternatives) == 1:
+            return alternatives[0]
+        return alternatives[self.id % 2]
+
+    def preferred_research_cutoff(self, alternatives):
+        return self._modulo_two_choice(alternatives)
+
+    def preferred_colonization_portion(self, alternatives):
+        return self._modulo_two_choice(alternatives)
+
+    def preferred_outpost_portion(self, alternatives):
+        return self._modulo_two_choice(alternatives)
+
+    def preferred_building_ratio(self, alternatives):
+        if not alternatives:
+            return None
+        return alternatives[self.id % 3] if len(alternatives) >= 3 else alternatives[self.id % len(alternatives)]
+
+    def preferred_discount_multiplier(self, alternatives):
+        return self._modulo_two_choice(alternatives)
+
+    def may_travel_beyond_supply(self, distance):
+        return (distance < 2
+                or distance <= 2 and self.aggression >= fo.aggression.typical
+                or self.aggression >= fo.aggression.aggressive)
+
+    # TODO remove this function as soon as old style research is gone
+    # It defeats the orthogonality goal of character components for little gain
+    def get_research_index(self):
+        research_index = self.id % 2
+        if self.aggression >= fo.aggression.aggressive:
+            research_index = 2 + (self.id % 3)  # so indices [2,3,4]
+        elif self.aggression >= fo.aggression.typical:
+            research_index += 1
+        return research_index
+
+
+class Character(Trait):
+    """
+    A collection of behaviours.
+
+    For each query that Trait supports a Character queries
+    all of its own traits to determine if an action is permissible or prefered.
+    """
+
+    def __init__(self, traits):
+        self.traits = traits
+        if not all([isinstance(x, Trait) for x in traits]):
+            raise TypeError("All traits must be sub-classes of Trait")
+
+    def get_trait(self, type_of_trait):
+        """Return the requested trait or None"""
+        trait = [x for x in self.traits if isinstance(x, type_of_trait)]
+        return trait[0] if trait else Trait()
+
+
+# Complete the Character class by adding all of the combiners to combine the outputs of the
+# individual traits.  Character tries to combine results in the way most limiting to the AI
```

Comment:
Either define the functions within the class or add a short comment there indicating that the class definition is incomplete.